imap_processing/swapi/l1/swapi_l1.py

```diff
@@ -1 +1,520 @@
+            full_sweep_sci[f"{cem_prefix}2"],
+            full_sweep_sci[f"{cem_prefix}3"],
+            full_sweep_sci[f"{cem_prefix}4"],
+            full_sweep_sci[f"{cem_prefix}5"],
+        ),
+        axis=0,
+    )
+
+    # Next:
+    # Reshape data by CEM, number of sweeps and sequence counts.
+    # Therefore, the data shape is 6 x total_full_sweeps x 12
+    # Output looks like this:
+    # [
+    # [[ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # [ 1  2  3  4  5  6  7  8  9  10  11  12]
+    # [ 2  3  4  5  6  7  8  9  10  11  12  13]]
+
+    # [[ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # [ 1  2  3  4  5  6  7  8  9  10  11  12]
+    # [ 2  3  4  5  6  7  8  9  10  11  12  13]]
+
+    # [[ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # [ 1  2  3  4  5  6  7  8  9  10  11  12]
+    # [ 2  3  4  5  6  7  8  9  10  11  12  13]]
+
+    # [[ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # [ 1  2  3  4  5  6  7  8  9  10  11  12]
+    # [ 2  3  4  5  6  7  8  9  10  11  12  13]]
+
+    # [[ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # [ 1  2  3  4  5  6  7  8  9  10  11  12]
+    # [ 2  3  4  5  6  7  8  9  10  11  12  13]]
+
+    # [[ 0  1  2  3  4  5  6  7  8  9 10 11]
+    # [ 1  2  3  4  5  6  7  8  9  10  11  12]
+    # [ 2  3  4  5  6  7  8  9  10  11  12  13]]]
+    # In other word, we grouped each cem's
+    # data by full sweep.
+    current_cem_counts = current_cem_counts.reshape(6, total_full_sweeps, 12)
+
+    # Then, we go from above to
+    # to this final output:
+    # [
+    # [[0  0  0  0  0  0]
+    # [1  1  1  1  1  1]
+    # [2  2  2  2  2  2]
+    # [3  3  3  3  3  3]
+    # [4  4  4  4  4  4]
+    # [5  5  5  5  5  5]
+    # [6  6  6  6  6  6]
+    # [7  7  7  7  7  7]
+    # [8  8  8  8  8  8]
+    # [9  9  9  9  9  9]
+    # [10 10 10 10 10 10]
+    # [11 11 11 11 11 11]],
+    # [[1  1  1  1  1  1]
+    # [2  2  2  2  2  2]
+    # [3  3  3  3  3  3]
+    # ...
+    # [12  12  12  12  12  12]],
+    # [[2  2  2  2  2  2]
+    # [3  3  3  3  3  3]
+    # ...
+    # [13  13  13  13  13  13]]
+    # ]
+    all_cem_data = np.stack(current_cem_counts, axis=-1)
+    # This line just flatten the inner most array to
+    # (total_full_sweeps x 72)
+    all_cem_data = all_cem_data.reshape(total_full_sweeps, 72)
+    return all_cem_data
+
+
+def process_swapi_science(sci_dataset):
+    """Process SWAPI science data and create CDF file.
+
+    Parameters
+    ----------
+    dataset : xarray.Dataset
+        L0 data
+    """
+    # ====================================================
+    # Step 1: Filter full cycle data
+    # ====================================================
+    full_sweep_indices = get_indices_of_full_sweep(sci_dataset)
+
+    # Filter full sweep data using indices returned from above line
+    full_sweep_sci = sci_dataset.isel({"Epoch": full_sweep_indices})
+
+    # Find indices of bad sweep cycles
+    bad_data_indices = check_for_bad_data(full_sweep_sci)
+    if len(bad_data_indices) > 0:
```

Comment:
I do that below like you are thinking. Is that what you mean? Did you want me to change anything here?

Reply:
I was wondering if you wanted to do the filtering instead of just the indices in the other routine. But I think it is fine here.

A few other thoughts here though that might make it easier to follow.
* Return the good_indices from the helper function and log the bad stuff there, just keep this area for filtering the datasets, not also calculating other indices. Push this logic up.
* Then I think you can do `good_sweep_sci = full_sweep_sci.isel()` no matter what is returned, not needing the if/else.

Reply:
Yes. I tried that and that worked out well and it was easier to follow.